### PR TITLE
Some fixes around FirExpressionEvaluator

### PR DESCRIPTION
**Pull request by ivandev0**
A bunch of small refactoring around types in FirExpressionEvaluator. Main idea: change "dynamic" function resolve (when arguments' types decide the function that is called) to "static" one (the same way it is done for runtime).

[KT-87836](https://youtrack.jetbrains.com/issue/KT-87836)

### CODE REVIEW

File: compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt [hunk 1]
```diff
@@ -782,82 +791,70 @@ fun ConstantValueKind.toCompileTimeType(): CompileTimeType {
 }
 
 // Unary operators
-private fun evaluateUnary(arg: FirExpression, callableId: CallableId): Any? {
+private fun evaluateUnary(arg: FirExpression, argType: ConeKotlinType, callableId: CallableId): Any? {
     if (arg !is FirLiteralExpression || arg.value == null) return null
 
-    val opr = arg.kind.convertToGivenKind(arg.value) ?: return null
+    val compileTimeType = argType.toCompileTimeType() ?: return null
+    val opr = argType.toConstantValueKind()?.convertToGivenKind(arg.value) ?: arg.value as Any
```
Comment (ivandev0) [inline compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt:798]: I would love to drop this strange conversion `argType.toConstantValueKind()?.convertToGivenKind(arg.value)`, but I can't. There is the following AA test (see implicitCastToByte)
```
fun main() {
    val b: Byte = <expr>-1</expr>
}
``` 

Here "1" is resolved to `Byte`, but "unaryMinus" to `Int.unaryMinus`. I am not sure if this is a bug or expected behavior.

File: compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt [hunk 2]
```diff
@@ -574,12 +584,9 @@ object FirExpressionEvaluator {
...
 
-            val opr1Value = opr1.kind.convertToGivenKind(opr1.value)
-            val opr2Value = opr2.kind.convertToGivenKind(opr2.value)
-
             val result = when (equalityOperatorCall.operation) {
-                FirOperation.EQ -> opr1Value == opr2Value
-                FirOperation.NOT_EQ -> opr1Value != opr2Value
+                FirOperation.EQ -> opr1.value == opr2.value
+                FirOperation.NOT_EQ -> opr1.value != opr2.value
                 else -> NotConst
```
Comment (bnorm) [inline compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt]: This should be `return NotConst` like in `visitComparisonExpression`, right?

**Review by bnorm — CHANGES_REQUESTED** [review]

File: compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt (hunk 2 above)
Comment (ivandev0) [reply compiler/fir/providers/src/org/jetbrains/kotlin/fir/expressions/FirExpressionEvaluator.kt]: Yep. Thanks for noticing!

**Review by bnorm — APPROVED** [review]
/final

**Review by abelkov — APPROVED** [review]
/final

File: plugins/kapt/kapt-compiler/src/org/jetbrains/kotlin/kapt/stubs/KaptStubConverter.kt [hunk 3]
```diff
@@ -1751,6 +1751,22 @@ class KaptStubConverter(val kaptContext: KaptContextForStubGeneration, val gener
                 sb.append(specialFpValueNumerator(value)).append(" / 0.0")
                 treeMaker.Binary(Tag.DIV, treeMaker.Literal(specialFpValueNumerator(value)), treeMaker.Literal(0.0))
             }
+            is UByte -> {
```
Comment (andrey-mogilev) [inline plugins/kapt/kapt-compiler/src/org/jetbrains/kotlin/kapt/stubs/KaptStubConverter.kt:1754]: Please add tests with unsigned values (including values that are out of range for their signed counter-parts) to `plugins/kapt/kapt-compiler/testData/converter/primitiveTypes.kt`.

Make sure to not commit auto-generated `primitiveTypes.jtree.txt` and `primitiveTypes.direct.txt`, and instead just update the common `primitiveTypes.txt` expectation

**Review by andrey-mogilev — CHANGES_REQUESTED** [review]

**Review by andrey-mogilev — APPROVED** [review]
/final
